jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/AsyncContent.java

```diff
@@ -47,13 +47,55 @@ public class AsyncContent implements Content.Sink, Content.Source, Closeable
     private Runnable demandCallback;
     private long length = UNDETERMINED_LENGTH;
 
+    /**
+     * {@inheritDoc}
+     * <p>The write completes when the {@link Content.Chunk} returned by {@link #read()}
+     * that wraps {@code byteBuffer} is released.</p>
+     */
     @Override
     public void write(boolean last, ByteBuffer byteBuffer, Callback callback)
     {
-        write(Content.Chunk.from(byteBuffer, last, callback::succeeded), callback);
+        Content.Chunk chunk;
+        if (!byteBuffer.hasRemaining())
+            chunk = last ? Content.Chunk.EOF : Content.Chunk.EMPTY;
+        else
+            chunk = Content.Chunk.from(byteBuffer, last, callback::succeeded);
```

Comment:
ditto

Reply:
This one  is an exception: the callback must not be succeeded until the write is over, which happens when another thread releases the generated chunk after reading it.

So we cannot let `from` call the release callback if it decides to replace the chunk with EOF or EMPTY otherwise we would succeed the callback too early.